crypto/dh/dh_check.c

```diff
@@ -257,9 +257,27 @@ int ossl_dh_check_priv_key(const DH *dh, const BIGNUM *priv_key, int *ret)
     two_powN = BN_new();
     if (two_powN == NULL)
         return 0;
-    if (dh->params.q == NULL)
+
+    if (dh->params.q != NULL) {
```

Comment:
FIPS should care about the fips186-4 and named group types.. For both these cases q needs to be set. If you want to do this code I suggest leaving the code how it was for fips mode.. i.e return an error if q is NULL.
 For FIPS 186-4 If someone loads a key and doesnt set the q, seed and counter themselves then the validation MUST FAIL.

For the non fips case it may need to be more flexible. Although arguably FIPS 186-2 should also fail if it doesnt have these values set either.

Reply:
Have you read the patch, Shane? It does not change anything when FIPS_MODULE is defined.